pkg/health/server/prober.go

```diff
@@ -234,34 +241,90 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
 	}
 }
 
-func (p *prober) updateIcmpStatus() {
-	p.Lock()
-	defer p.Unlock()
+const httpPathDescription = "Via L3"
+
+func (p *prober) getIPsByNode() map[string][]*net.IPAddr {
+	p.RLock()
+	defer p.RUnlock()
 
-	for _, status := range p.results {
-		if status.Icmp == nil {
-			status.Icmp = &models.ConnectivityStatus{}
-			status.Icmp.Status = "Connection timed out"
+	// p.nodes is mapped from all known IPs -> nodes in N:M configuration,
+	// so multiple IPs could refer to the same node. To ensure we only
+	// ping each node once, deduplicate nodes into map of nodeName -> []IP.
+	nodes := make(map[string][]*net.IPAddr)
+	for _, node := range p.nodes {
+		if nodes[node.Name] != nil {
+			// Already handled this node.
+			continue
+		}
+		nodes[node.Name] = []*net.IPAddr{}
+		for elem, primary := range node.Addresses() {
+			if _, addr := resolveIP(&node, elem, primary); addr != nil {
+				nodes[node.Name] = append(nodes[node.Name], addr)
+			}
 		}
 	}
+
+	return nodes
 }
 
-func (p *prober) clearIcmpStatus() {
-	p.Lock()
-	defer p.Unlock()
+func icmpPing(node string, ip string, ctx context.Context, resChan chan<- connectivityResult, wg *sync.WaitGroup, probeDeadline time.Duration) {
+	defer wg.Done()
 
-	for _, status := range p.results {
-		status.Icmp = nil
+	result := &models.ConnectivityStatus{}
+
+	// Only add fields to the scoped logger if debug is enabled, to save on resources.
+	// This can be done since all logs in this function are debug-level only.
+	scopedLog := log
+	debugLogsEnabled := logging.CanLogAt(scopedLog.Logger, logrus.DebugLevel)
+	if debugLogsEnabled {
+		scopedLog = log.WithFields(logrus.Fields{
+			logfields.NodeName: node,
+			logfields.IPAddr:   ip,
+		})
 	}
-}
 
-const httpPathDescription = "Via L3"
+	scopedLog.Debug("Pinging host")
+	pinger, err := probing.NewPinger(ip)
+	if err != nil {
+		if debugLogsEnabled {
+			scopedLog.WithError(err).Debug("Failed to create pinger")
+		}
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
 
-func (p *prober) httpProbe(node string, ip string) *models.ConnectivityStatus {
-	result := &models.ConnectivityStatus{}
+	pinger.Timeout = probeDeadline
+	pinger.Count = 1
+	pinger.OnFinish = func(stats *probing.Statistics) {
+		if stats.PacketsRecv > 0 && len(stats.Rtts) > 0 {
+			if debugLogsEnabled {
+				scopedLog.WithField("rtt", stats.Rtts[0].Nanoseconds()).Debug("Ping successful")
+			}
+			result.Latency = stats.Rtts[0].Nanoseconds()
+		} else {
+			scopedLog.Debug("Ping failed")
+			result.Status = "Connection timed out"
+		}
+	}
+	pinger.SetPrivileged(true)
+	err = pinger.Run()
```

Comment:
nit: we could use `pinger.RunWithContext` with context that has timeout of `probeDeadline` to make sure it always finishes within timeout :)